platform/posix/transport/src/mbedtls_pkcs11_posix.c

```diff
@@ -0,0 +1,958 @@
+        }
+    }
+
+    /* Map the mbedTLS algorithm to its internal metadata. */
+    if( ret == CKR_OK )
+    {
+        memcpy( &pCtx->privKeyInfo, mbedtls_pk_info_from_type( keyAlgo ), sizeof( mbedtls_pk_info_t ) );
+
+        pCtx->privKeyInfo.sign_func = privateKeySigningCallback;
+        pCtx->privKey.pk_info = &pCtx->privKeyInfo;
+        pCtx->privKey.pk_ctx = pCtx;
+    }
+
+    return ret;
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t privateKeySigningCallback( void * pContext,
+                                          mbedtls_md_type_t mdAlg,
+                                          const unsigned char * pHash,
+                                          size_t hashLen,
+                                          unsigned char * pSig,
+                                          size_t * pSigLen,
+                                          int32_t ( *pRng )( void *,
+                                                             unsigned char *,
+                                                             size_t ),
+                                          void * pRngContext )
+{
+    CK_RV ret = CKR_OK;
+    int32_t result = 0;
+    MbedtlsPkcs11Context_t * pMbedtlsPkcs11Context = ( MbedtlsPkcs11Context_t * ) pContext;
+    CK_MECHANISM mech = { 0 };
+    CK_BYTE toBeSigned[ 256 ];
+    CK_ULONG toBeSignedLen = sizeof( toBeSigned );
+
+    /* Unreferenced parameters. */
+    ( void ) ( pRng );
+    ( void ) ( pRngContext );
+    ( void ) ( mdAlg );
+
+    /* Sanity check buffer length. */
+    if( hashLen > sizeof( toBeSigned ) )
+    {
+        ret = CKR_ARGUMENTS_BAD;
+    }
+
+    /* Format the hash data to be signed. */
+    if( CKK_RSA == pMbedtlsPkcs11Context->keyType )
+    {
+        mech.mechanism = CKM_RSA_PKCS;
+
+        /* mbedTLS expects hashed data without padding, but PKCS #11 C_Sign function performs a hash
+         * & sign if hash algorithm is specified.  This helper function applies padding
+         * indicating data was hashed with SHA-256 while still allowing pre-hashed data to
+         * be provided. */
+        ret = vAppendSHA256AlgorithmIdentifierSequence( ( const uint8_t * ) pHash, toBeSigned );
+        toBeSignedLen = pkcs11RSA_SIGNATURE_INPUT_LENGTH;
+    }
+    else if( CKK_EC == pMbedtlsPkcs11Context->keyType )
+    {
+        mech.mechanism = CKM_ECDSA;
+        memcpy( toBeSigned, pHash, hashLen );
+        toBeSignedLen = hashLen;
+    }
+    else
+    {
+        ret = CKR_ARGUMENTS_BAD;
+    }
+
+    if( CKR_OK == ret )
+    {
+        /* Use the PKCS#11 module to sign. */
+        ret = pMbedtlsPkcs11Context->pP11FunctionList->C_SignInit( pMbedtlsPkcs11Context->p11Session,
+                                                                   &mech,
+                                                                   pMbedtlsPkcs11Context->p11PrivateKey );
+    }
+
+    if( CKR_OK == ret )
+    {
+        *pSigLen = sizeof( toBeSigned );
```

Comment:
This assignment seems unnecessary as `C_Sign` will update the value with the signature length?

Reply:
It is needed.